examples/llm_inference/hugging_face_js/gemma_3n/index.js

```diff
@@ -0,0 +1,387 @@
+  };
+  // clear old oauth, if any
+  localStorage.removeItem("oauth");
+}
+
+/**
+ * Utility function to show progress while we load from remote file into local
+ * cache.
+ */
+async function pipeStreamAndReportProgress(readableStream, writableStream) {
+  // Effectively "await responseStream.pipeTo(writeStream)", but with progress
+  // reporting.
+  const reader = readableStream.getReader();
+  const writer = writableStream.getWriter();
+  let bytesCount = 0;
+  let progressBarPercent = 0;
+  let wasAborted = false;
+  try {
+    while (true) {
+      const {done, value} = await reader.read();
+      if (done) {
+        break;
+      }
+      if (value) {
+        bytesCount += value.length;
+        const percentage = Math.round(bytesCount / modelSize * 90);
+        if (percentage > progressBarPercent) {
+          progressBarPercent = percentage;
+          updateProgressBar(progressBarPercent);
+        }
+        await writer.write(value);
+      }
+    }
+  } catch (error) {
+    console.error('Error while piping stream:', error);
+    // Abort the writer if there's an error
+    wasAborted = true;
+    await writer.abort(error);
+    throw error;
+  } finally {
+    // Release the reader lock
+    reader.releaseLock();
+    // Close the writer only if the stream wasn't aborted
+    if (!wasAborted) {
+      console.log('Closing the writer, and hence the stream');
+      await writer.close();
+    }
+  }
+}
+
+/**
+ * Loads the LLM file from either cache or OAuth-guarded remote download.
+ */
+async function loadLlm() {
+  let opfs = await navigator.storage.getDirectory();
+  // If we can load the model from cache, then do so.
+  try {
+    const fileHandle = await opfs.getFileHandle(cacheFileName);
+    // Check to make sure size is as expected, and not a partially-downloaded
+    // or corrupt file.
+    console.log('Model found in cache; checking size.');
+    const file = await fileHandle.getFile();
+    console.log('File size is: ', file.size);
+    if (file.size !== modelSize) {
+      console.error('Cached model had unexpected size. Redownloading.');
+      throw new Error('Unexpected cached model size');
+    }
+    console.log('Model found in cache of expected size, reusing.');
+    const fileReader = file.stream().getReader();
+    await initLlm(fileReader);
+  } catch {
+    // Otherwise, we need to be download remotely, which requires oauth.
+    console.log('Model not found in cache: oauth and download required.');
+    // We first remove from cache, in case model file is corrupted/partial.
+    try {
+      await opfs.removeEntry(cacheFileName);
+    } catch {}
+    let oauthResult = localStorage.getItem("oauth");
+    if (oauthResult) {
+      try {
+        oauthResult = JSON.parse(oauthResult);
+      } catch {
+        oauthResult = null;
+      }
+    }
+    oauthResult ||= await oauthHandleRedirectIfPresent();
```

Comment:
Nit: Should this be `??=`?

Reply:
This was taken directly from the official HF oauth example, so would prefer to leave as-is for clarity.